### PR TITLE
osutil/kcmdline: provide marshalers for ArgumentPattern

### PR DESCRIPTION
**Pull request by valentindavid**
Needed for https://github.com/canonical/ubuntu-image/pull/136

### CODE REVIEW

File: osutil/kcmdline/kcmdline.go [hunk 1]
```diff
@@ -441,3 +445,39 @@ func (kap *ArgumentPattern) UnmarshalYAML(unmarshal func(interface{}) error) err
 
 	return nil
 }
+
+func (kap *ArgumentPattern) marshalToString() (string, error) {
+	return fmt.Sprintf(`%s=%s`, kap.param, kap.value.String()), nil
```
Comment (MiguelPires) [inline osutil/kcmdline/kcmdline.go:450]: The error can be removed from the return type
```suggestion
func (kap *ArgumentPattern) marshalToString() string {
	return fmt.Sprintf(`%s=%s`, kap.param, kap.value.String())
```

File: osutil/kcmdline/kcmdline.go [hunk 2]
```diff
@@ -441,3 +445,39 @@ func (kap *ArgumentPattern) UnmarshalYAML(unmarshal func(interface{}) error) err
...
+func (kap *ArgumentPattern) UnmarshalBinary(data []byte) error {
+	return kap.unmarshalFromString(string(data))
+}
+
+func (kap *ArgumentPattern) MarshalBinary() ([]byte, error) {
+	str, err := kap.marshalToString()
+	if err != nil {
+		return []byte{}, err
+	}
+	return []byte(str), nil
```
Comment (MiguelPires) [inline osutil/kcmdline/kcmdline.go:462]: Then these can be slightly simplified
```suggestion
	return []byte(kap.marshalToString()), nil
```

File: osutil/kcmdline/kcmdline.go [hunk 3]
```diff
@@ -441,3 +445,39 @@ func (kap *ArgumentPattern) UnmarshalYAML(unmarshal func(interface{}) error) err
...
+func (kap *ArgumentPattern) UnmarshalText(data []byte) error {
+	return kap.unmarshalFromString(string(data))
+}
+
+func (kap *ArgumentPattern) MarshalText() ([]byte, error) {
+	str, err := kap.marshalToString()
+	if err != nil {
+		return []byte{}, err
+	}
+	return []byte(str), nil
```
Comment (MiguelPires) [inline osutil/kcmdline/kcmdline.go:474]:
```suggestion
	return []byte(kap.marshalToString()), nil
```

File: osutil/kcmdline/kcmdline.go [hunk 4]
```diff
@@ -441,3 +445,39 @@ func (kap *ArgumentPattern) UnmarshalYAML(unmarshal func(interface{}) error) err
...
+}
+
+func (kap *ArgumentPattern) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	var arg string
+	if err := unmarshal(&arg); err != nil {
+		return errors.New("cannot unmarshal kernel argument")
```
Comment (MiguelPires) [inline osutil/kcmdline/kcmdline.go:480]: Should this include the underlying error?
```suggestion
		return fmt.Errorf("cannot unmarshal kernel argument: %v", err)
```

File: osutil/kcmdline/kcmdline.go (hunk 1 above)
Comment (valentindavid) [reply osutil/kcmdline/kcmdline.go:450]: I should have re-read my code. I had something else in mind at first. Will fix that.

File: osutil/kcmdline/kcmdline.go (hunk 3 above)
Comment (valentindavid) [reply osutil/kcmdline/kcmdline.go:474]: With the other fix, this will be simplified.

File: osutil/kcmdline/kcmdline.go (hunk 4 above)
Comment (valentindavid) [reply osutil/kcmdline/kcmdline.go:480]: Good point.

Comment (valentindavid) [reply osutil/kcmdline/kcmdline.go]: I had forgot to fix this one. Now it is pushed.

**Review by Meulengracht — APPROVED** [review]

**Pull request merged**